DTO to EMF converters

ideas-into-software opened the pull request:
Continuing from https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/54 - branch `issues-51-52-DTO-to-EMF` re-created to resolve issues with commits as requested in https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/54#issuecomment-2056682822 ("Change Requests" -> "1").

PR contains both "DTO to EPackage converter" ( https://github.com/geckoprojects-org/org.gecko.emf.utils/issues/51 ) and "DTO to EObject converter" ( https://github.com/geckoprojects-org/org.gecko.emf.utils/issues/52 ), in addition to change requests discussed in https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/54, i.e.: 

 - Externalize custom `EDataType`s ( `org.gecko.emf.converter.model` );
 
 - Dynamically register EMF built-in and custom `EDataType`s ( static block in `org.gecko.emf.converter.DTOToEPackageConverter` );
 
 - Explicitly register `EPackage` ( to make it visible that registation is explicit, `EPackage` registration was moved to test cases in `org.gecko.emf.converter.tests.DTOToEPackageConverterTest` and `org.gecko.emf.converter.tests.DTOToEObjectConverterTest` );
 
 - For util classes, use class with static methods instead of enum type;

 - Fresh checkout of https://github.com/geckoprojects-org/org.gecko.emf.utils `snapshot` to resolve issues with commits;
 

### Questions (c.d.)

Following is list of questions / open points which need to be clarified in order to close this project: 

 1) Regarding "DTO to EObject converter": Use custom `org.osgi.util.converter.Converter` implementation or standard (existing) `org.osgi.util.converter.Converter` implementation, with custom implementation of `org.osgi.util.converter.ConverterFunction` and `org.osgi.util.converter.TypeRule` ? See https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/54#issuecomment-2076021591 for more details;
  
 2) Should `DTOToEPackageConverter` be in the same bundle as `DTOToEObjectConverter*` classes or moved to separate bundles ? If separate bundles, any naming preference ? ( e.g. `org.gecko.emf.converter.dto_to_eobject`, `org.gecko.emf.converter.dto_to_epackage` );
 
 3) Regarding relaxing requirement of accepting only classes which extend `org.osgi.dto.DTO`, as discussed in https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/54#issuecomment-2056682822 - should only `org.gecko.emf.converter.DTOToEPackageConverter.convert(String, String, String, Class<? extends DTO>...)` method signature be modified ? Currently, whether given class is a DTO is checked also via `org.gecko.emf.converter.DTOUtil.isDTOType(Class<?>, boolean)` - should such check be also removed ?

ideas-into-software committed:
DTO to EMF converters

 - DTO to EPackge converter ( #51 )

 - Converter Impl for DTO to EObject ( #52 ) - skeleton

Signed-off-by: Michael H. Siemaszko <mhs@into.software>

ideas-into-software committed:
DTO to EMF converters (c.d.)

 - Converter Impl for DTO to EObject ( #52 )

Signed-off-by: Michael H. Siemaszko <mhs@into.software>

ideas-into-software committed:
DTO to EMF converters (c.d.)

- Externalize custom EDataTypes
 
 - Dynamically register EMF built-in and custom EDataTypes
 
 - Explicitly register EPackage
 
 - Minor fixes and improvements


Signed-off-by: Michael H. Siemaszko <mhs@into.software>

juergen-albert commented:
> Regarding "DTO to EObject converter": Use custom org.osgi.util.converter.Converter implementation or standard (existing) org.osgi.util.converter.Converter implementation, with custom implementation of org.osgi.util.converter.ConverterFunction and org.osgi.util.converter.TypeRule ? See https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/54#issuecomment-2076021591 for more details;

Good Question and I have to admit I'm not sure. The Specification only speaks about using the Converter Builder and feed it with custom rules and functions.  If your functionality can be covered with that, I thing you should stick to the Specification and should not create a custom Converter Implementation and only provide some method to get a preconfigured Converter and ConverterBuilder.

> Should DTOToEPackageConverter be in the same bundle as DTOToEObjectConverter* classes or moved to separate bundles ? If separate bundles, any naming preference ? ( e.g. org.gecko.emf.converter.dto_to_eobject, org.gecko.emf.converter.dto_to_epackage );

It can remain in the same Bundle.

> Regarding relaxing requirement of accepting only classes which extend org.osgi.dto.DTO, as discussed in https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/54#issuecomment-2056682822 - should only org.gecko.emf.converter.DTOToEPackageConverter.convert(String, String, String, Class<? extends DTO>...) method signature be modified ? Currently, whether given class is a DTO is checked also via org.gecko.emf.converter.DTOUtil.isDTOType(Class<?>, boolean) - should such check be also removed ?

I would not remove the method completely, as it makes sense even if we want to convert Classes that are no DTOs. For now, I would suggest allowing classes that have static methods as well. Non static methods may rely on non fields, which will lead to a useless model.

juergen-albert commented:
The commit itself looks good so far. One question comes to mind though: What is the rational behind the `org.osgi.util.converter-1.1.0.jar` in the local index repo?

ideas-into-software commented:
@juergen-albert

Please see PR https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/56, which brings work on converters to mergeable state, as per your preferences, i.e.: 

 - Refactored to use standard OSGi converter with custom `org.osgi.util.converter.ConverterFunction` and `org.osgi.util.converter.TypeRule`
 
 - Removed refactored OSGi converter implementation from local repository
 
 - Relaxed requirement of accepting only classes which extend `org.osgi.dto.DTO`
 
 
To answer your question from https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/55#issuecomment-2106350769: 

 * I partially answered this question via https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/54#issuecomment-2076021591, posted on 2024/04/25;

 * It is a refactored OSGi converter implementation, open for extension; this was needed for the "custom `org.osgi.util.converter.Converter` implementation", mentioned numerous times throughout this and previous PR ( https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/54 );

 * We talked about "custom implementation of `org.osgi.util.converter.Converter`", even requirement which kicked this off ( https://github.com/geckoprojects-org/org.gecko.emf.utils/issues/52 ) was titled "Converter Impl for DTO to EObject" - that is how I understood "custom implementation", i.e. to create class implementing `org.osgi.util.converter.Converter` interface, which handles such DTO to EMF conversions; as I started working on it, it turned out it is closed for extension, including the `org.osgi.util.converter.InternalConverter` interface which it uses underneath and all other were package-private classes;
 
 * I wanted to avoid duplicating 1000+ lines of code - my understanding was that this custom DTO to EMF converter is supposed to handle all other conversions as well, like standard converter already does; such are also used in the process of converting DTO to EMF, so without such a separate instance of standard converter would need to be used underneath.
 
 * I did not know its internals previously, neither used it that much ( except for `org.osgi.util.converter.Converter.convert(Object)`), especially `org.osgi.util.converter.Converter.function()` and `org.osgi.util.converter.Converter.newConverterBuilder()`; as I got to know it better, I found out I can use only custom implementation of `org.osgi.util.converter.ConverterFunction` and custom implementation of `org.osgi.util.converter.TypeRule` with standard (existing) `org.osgi.util.converter.Converter`; that is why I provided both solutions, one based on "custom `org.osgi.util.converter.Converter` implementation" and one based on standard (existing) `org.osgi.util.converter.Converter`;
 
Since this was obviously a misunderstanding - and by this I mean the "custom implementation" / "Converter Impl" part, which you mentioned from the very start and which got me thinking along these lines - I replaced this PR ( https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/55 ) with another PR ( https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/56 ), which does not contain `org.osgi.util.converter-1.1.0.jar` ( i.e. refactored OSGi converter implementation so it's open for extension ), with code that is refactored to use only standard (existing) `org.osgi.util.converter.Converter`.

If you would like to have that refactored OSGi converter implementation shared as well, I can submit a separate PR to OSGi project directly; the JAR ( https://github.com/geckoprojects-org/org.gecko.emf.utils/blob/issues-51-52-DTO-to-EMF/cnf/local/org.osgi.util.converter/org.osgi.util.converter-1.1.0.jar ) from local repo in previous PR contains source code as well, in case you want to check what changes were introduced in that refactored `org.osgi.util.converter`.
